application-analytics-default/src/main/java/com/xwiki/analytics/internal/MostViewedJsonNormaliser.java

```diff
@@ -0,0 +1,200 @@
+import org.xwiki.resource.CreateResourceTypeException;
+import org.xwiki.resource.ResourceReference;
+import org.xwiki.resource.ResourceReferenceResolver;
+import org.xwiki.resource.ResourceType;
+import org.xwiki.resource.ResourceTypeResolver;
+import org.xwiki.resource.UnsupportedResourceReferenceException;
+import org.xwiki.resource.entity.EntityResourceReference;
+import org.xwiki.url.ExtendedURL;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.xwiki.analytics.JsonNormaliser;
+
+import liquibase.repackaged.org.apache.commons.lang3.exception.ExceptionUtils;
+
+/**
+ * The normaliser for the MostViewedMacro.
+ *
+ * @version $Id$
+ * @since 1.0
+ */
+@Component
+@Named("MostViewedPages")
+@Singleton
+public class MostViewedJsonNormaliser implements JsonNormaliser
+{
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+
+    private static final String DATE = "date";
+
+    private static final String URL = "url";
+
+    private static final String LABEL = "label";
+
+    @Inject
+    private ResourceReferenceResolver<ExtendedURL> resourceReferenceResolver;
+
+    @Inject
+    private Logger logger;
+
+    @Inject
+    private ResourceTypeResolver<ExtendedURL> resourceTypeResolver;
+
+    @Inject
+    @Named("compactwiki")
+    private EntityReferenceSerializer<String> serializer;
+
+    /**
+     * This method will normalise the jsons returned by Matomo into a single format.
+     *
+     * @param jsonString the json provided by Matomo.
+     * @return the normalised json as a string.
+     */
+    public JsonNode normaliseData(String jsonString) throws JsonProcessingException
+    {
+        // I need to convert the string returned by matomo in a JSON to easily handle the processing of the nodes.
+        JsonNode jsonRoot = OBJECT_MAPPER.readTree(jsonString);
+        // Matomo may return several variants of JSON formats. In one scenario, when the period is set to
+        // day/week/month/year, it returns a JSON object with keys representing dates. The corresponding value for
+        // each key is an array of JSON objects, each of which represents a page. However, if the user sets the
+        // period parameter to "range" Matomo returns an array of JSON objects, with each JSON object representing
+        // a page. Due to these variations, I need to process the result from Matomo to create a normalized format.
+        // This normalized format is an array of JSON objects and each JSON object in this array will have a new
+        // field called 'date'. This 'date' field will be set to 'N/A' when Matomo returns an array instead of an
+        // object. In the 'processArrayNode' and 'processObjectNode' methods, I also modify the 'label' field to
```

Comment:
`In the 'processArrayNode' and 'processObjectNode' methods, I also...` -> `For both type of formats, the label field is also altered in order to contain the full page name`
But I think this last phrase could be added directly to each of the 2 methods description, since you need now to specify it there as well